apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Trace/SimpleManagedTracer.cs

```diff
@@ -168,16 +178,21 @@ public T RunInSpan<T>(Func<T> func, string name, StartSpanOptions options = null
         /// <inheritdoc />
         public void EndSpan()
         {
-            lock (_stackLock)
-            {
-                CheckStackNotEmpty();
+            var currentStack = TraceStack;
+            CheckStackNotEmpty(currentStack);
 
-                TraceSpan span = _traceStack.Pop();
-                span.EndTime = Timestamp.FromDateTime(DateTime.Now.ToUniversalTime());
+            TraceSpan span;
+            currentStack = currentStack.Pop(out span);
+            TraceStack = currentStack;
+            span.EndTime = Timestamp.FromDateTime(DateTime.Now.ToUniversalTime());
 
+            lock (_traceLock)
+            {
                 _trace.Spans.Add(span);
 
-                if (_traceStack.Count == 0)
+                var newOpenSpanCount = Interlocked.Decrement(ref _openSpanCount);
+                Debug.Assert(newOpenSpanCount >= 0, "Invalid open span count");
```

Comment:
This is really a "this should never happen" kind of thing. I don't know if we need the runtime overhead, although small, of checking on every span. Also, I'm a little worried about the debuggability of something like this: if you call end span too many times, you won't know until one of the last open spans are closed, which could happen later and on a different thread.

Personally, I think it would be safer to hide EndSpan and force users to dispose the result of StartSpan, and then perhaps have that disposable object do a GaxPreconditions.CheckState to ensure it is never double-disposed. But that should be for another PR.

If you'd still like me to change this though, let me know and I will.